crates/anvil-polkadot/src/substrate_node/lazy_loading/backend.rs

```diff
@@ -0,0 +1,1855 @@
+                let readable_db = self.db.read();
+                return Ok(readable_db.storage(key).ok().flatten());
+            }
+        }
+
+        let readable_db = self.db.read();
+        let maybe_storage = readable_db.storage(key);
+        let value = match maybe_storage {
+            Ok(Some(data)) => Some(data),
+            _ if !self.removed_keys.read().contains_key(key) => {
+                // Only try remote fetch if RPC client is available
+                let result =
+                    if self.rpc().is_some() { remote_fetch(Some(self.fork_block)) } else { None };
+
+                // Cache state
+                drop(readable_db);
+                self.update_storage(key, &result);
+
+                result
+            }
+            _ => None,
+        };
+
+        Ok(value)
+    }
+
+    fn storage_hash(
+        &self,
+        key: &[u8],
+    ) -> Result<Option<<HashingFor<Block> as sp_core::Hasher>::Out>, Self::Error> {
+        let remote_fetch = |block: Option<Block::Hash>| -> Result<
+            Option<<HashingFor<Block> as sp_core::Hasher>::Out>,
+            Self::Error,
+        > {
+            match self.rpc() {
+                Some(rpc) => rpc
+                    .storage_hash(StorageKey(key.to_vec()), block)
+                    .map_err(|e| format!("Failed to fetch storage hash from RPC: {e:?}")),
+                None => Ok(None),
+            }
+        };
+
+        // When before_fork, try RPC first, then fall back to local DB
+        if self.before_fork {
+            if self.rpc().is_some() {
+                return remote_fetch(self.block_hash);
+            } else {
+                // No RPC client, try to read from local DB
+                return Ok(self.db.read().storage_hash(key).ok().flatten());
+            }
+        }
+
+        let storage_hash = self.db.read().storage_hash(key);
+        match storage_hash {
+            Ok(Some(hash)) => Ok(Some(hash)),
+            _ if !self.removed_keys.read().contains_key(key) => {
+                if self.rpc().is_some() {
+                    remote_fetch(Some(self.fork_block))
+                } else {
+                    Ok(None)
+                }
+            }
+            _ => Ok(None),
+        }
+    }
+
+    fn closest_merkle_value(
+        &self,
+        _key: &[u8],
+    ) -> Result<
+        Option<sp_trie::MerkleValue<<HashingFor<Block> as sp_core::Hasher>::Out>>,
+        Self::Error,
+    > {
+        unimplemented!("closest_merkle_value: unsupported feature for lazy loading")
+    }
+
+    fn child_closest_merkle_value(
+        &self,
+        _child_info: &sp_storage::ChildInfo,
+        _key: &[u8],
+    ) -> Result<
+        Option<sp_trie::MerkleValue<<HashingFor<Block> as sp_core::Hasher>::Out>>,
+        Self::Error,
+    > {
+        unimplemented!("child_closest_merkle_value: unsupported feature for lazy loading")
+    }
+
+    fn child_storage(
```

Comment:
the child storage functions are needed, because smart contracts store their state as subtries